From 22fa7a869995df418d9f108e60cd913e781ede0a Mon Sep 17 00:00:00 2001
From: Wei-Luan Wang <weiluanwang@google.com>
Date: Mon, 27 Oct 2025 03:07:33 +0000
Subject: [PATCH] Revert "Remove Do* Methods in MockObjectProxy"

This reverts most of commit 42d69598b9e505798a55a8b569ce4953afb3e3ea.

It does however create mock functions for the "proper" methods so that
we can migrate client packages. To not break current code, the
MockObjectProxy sets default handlers for all mock methods that call the
Do* functions.

patch-name:
backward-compatibility-3300-Revert-Remove-Do-Methods-in-MockObjectProxy.patch
Change-Id: Ic1898f92a264ba44aa88e84f07770c8a3d292ba6

Change-Id: I895ed2807a75d05e4ba70e1dc082b48cbea6dadd
---
 dbus/mock_object_proxy.cc | 27 +++++++++++++++++++++++++++
 dbus/mock_object_proxy.h  | 29 ++++++++++++++++++++++++++++-
 dbus/mock_unittest.cc     |  6 +++---
 3 files changed, 58 insertions(+), 4 deletions(-)

diff --git a/dbus/mock_object_proxy.cc b/dbus/mock_object_proxy.cc
index 81c537eccc..254802aa28 100644
--- a/dbus/mock_object_proxy.cc
+++ b/dbus/mock_object_proxy.cc
@@ -4,12 +4,39 @@
 
 #include "dbus/mock_object_proxy.h"
 
+using ::testing::_;
+
 namespace dbus {
 
 MockObjectProxy::MockObjectProxy(Bus* bus,
                                  const std::string& service_name,
                                  const ObjectPath& object_path)
     : ObjectProxy(bus, service_name, object_path, DEFAULT_OPTIONS) {
+  ON_CALL(*this, CallMethod(_, _, _)).WillByDefault(
+      [this](MethodCall* method_call, int timeout_ms,
+             ResponseCallback callback) {
+        DoCallMethod(method_call, timeout_ms, &callback);
+      }
+  );
+  ON_CALL(*this, CallMethodWithErrorResponse(_, _, _)).WillByDefault(
+      [this](MethodCall* method_call, int timeout_ms,
+             ResponseOrErrorCallback callback) {
+        DoCallMethodWithErrorResponse(method_call, timeout_ms, &callback);
+      }
+  );
+  ON_CALL(*this, WaitForServiceToBeAvailable(_)).WillByDefault(
+      [this](WaitForServiceToBeAvailableCallback callback) {
+        DoWaitForServiceToBeAvailable(&callback);
+      }
+  );
+  ON_CALL(*this, ConnectToSignal(_, _, _, _)).WillByDefault(
+      [this](const std::string& interface_name, const std::string& signal_name,
+             SignalCallback signal_callback,
+             OnConnectedCallback on_connected_callback) {
+        DoConnectToSignal(interface_name, signal_name, signal_callback,
+                          &on_connected_callback);
+      }
+  );
 }
 
 MockObjectProxy::~MockObjectProxy() = default;
diff --git a/dbus/mock_object_proxy.h b/dbus/mock_object_proxy.h
index 37b7e26301..3ac07745ec 100644
--- a/dbus/mock_object_proxy.h
+++ b/dbus/mock_object_proxy.h
@@ -29,24 +29,51 @@ class MockObjectProxy : public ObjectProxy {
       base::expected<std::unique_ptr<Response>, Error>(MethodCall* method_call,
                                                        int timeout_ms));
 
+  // This method is not mockable because it takes a move-only argument. To work
+  // around this, CallMethod() implementation here calls DoCallMethod() which is
+  // mockable.
   MOCK_METHOD3(CallMethod,
+               void(MethodCall* method_call,
+                  int timeout_ms,
+                  ResponseCallback callback));
+  MOCK_METHOD3(DoCallMethod,
                void(MethodCall* method_call,
                     int timeout_ms,
-                    ResponseCallback callback));
+                    ResponseCallback* callback));
 
+  // This method is not mockable because it takes a move-only argument. To work
+  // around this, CallMethodWithErrorResponse() implementation here calls
+  // DoCallMethodWithErrorResponse() which is mockable.
   MOCK_METHOD3(CallMethodWithErrorResponse,
                void(MethodCall* method_call,
                     int timeout_ms,
                     ResponseOrErrorCallback callback));
+  MOCK_METHOD3(DoCallMethodWithErrorResponse,
+               void(MethodCall* method_call,
+                    int timeout_ms,
+                    ResponseOrErrorCallback* callback));
 
+  // This method is not mockable because it takes a move-only argument. To work
+  // around this, WaitForServiceToBeAvailable() implementation here calls
+  // DoWaitForServiceToBeAvailable() which is mockable.
   MOCK_METHOD1(WaitForServiceToBeAvailable,
                void(WaitForServiceToBeAvailableCallback callback));
+  MOCK_METHOD1(DoWaitForServiceToBeAvailable,
+               void(WaitForServiceToBeAvailableCallback* callback));
 
+  // This method is not mockable because it takes a move-only argument. To work
+  // around this, ConnectToSignal() implementation here calls
+  // DoConnectToSignal() which is mockable.
   MOCK_METHOD4(ConnectToSignal,
                void(const std::string& interface_name,
                     const std::string& signal_name,
                     SignalCallback signal_callback,
                     OnConnectedCallback on_connected_callback));
+  MOCK_METHOD4(DoConnectToSignal,
+               void(const std::string& interface_name,
+                    const std::string& signal_name,
+                    SignalCallback signal_callback,
+                    OnConnectedCallback* on_connected_callback));
   MOCK_METHOD1(SetNameOwnerChangedCallback,
                void(NameOwnerChangedCallback callback));
   MOCK_METHOD0(Detach, void());
diff --git a/dbus/mock_unittest.cc b/dbus/mock_unittest.cc
index de4ed6dc7b..87bea1b727 100644
--- a/dbus/mock_unittest.cc
+++ b/dbus/mock_unittest.cc
@@ -50,7 +50,7 @@ class MockTest : public testing::Test {
 
     // Set an expectation so mock_proxy's CallMethod() will use
     // HandleMockProxyResponseWithMessageLoop() to return responses.
-    EXPECT_CALL(*mock_proxy_.get(), CallMethod(_, _, _))
+    EXPECT_CALL(*mock_proxy_.get(), DoCallMethod(_, _, _))
         .WillRepeatedly(
             Invoke(this, &MockTest::HandleMockProxyResponseWithMessageLoop));
 
@@ -114,13 +114,13 @@ class MockTest : public testing::Test {
   void HandleMockProxyResponseWithMessageLoop(
       MethodCall* method_call,
       int timeout_ms,
-      ObjectProxy::ResponseCallback response_callback) {
+      ObjectProxy::ResponseCallback* response_callback) {
     std::unique_ptr<Response> response =
         CreateMockProxyResponse(method_call, timeout_ms).value_or(nullptr);
     task_environment_.GetMainThreadTaskRunner()->PostTask(
         FROM_HERE,
         base::BindOnce(&MockTest::RunResponseCallback, base::Unretained(this),
-                       std::move(response_callback), std::move(response)));
+                       std::move(*response_callback), std::move(response)));
   }
 
   // Runs the given response callback with the given response.
-- 
2.52.0.rc1.455.g30608eb744-goog

